### ISSUE TITLE
100% CPU on "Resolving dependencies through SAT"

### ISSUE THREAD

**Issue by williamdes** [composer/composer#10268]
My `composer.json`:

https://github.com/code-lts/Laravel-FCM/blob/1.6.2/composer.json

Output of `composer diagnose`:

```
Checking composer.json: OK
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 2.1.12
PHP version: 7.4.24
PHP binary path: /usr/bin/php7.4
OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
cURL version: 7.74.0 libz 1.2.11 ssl OpenSSL/1.1.1j
zip: extension present, unzip present, 7-Zip present (7z)
```

When I run this command (with no vendor folder or lock file): composer update

```
Updating dependencies
Generating rules
Resolving dependencies through SAT
```

And I expected this to happen: Work correctly like it does using v1 of composer (just tried it before posting).

I just upgraded in case it would help but no.
```
Upgrading to version 2.1.12 (stable channel).
Use composer self-update --rollback to return to version 2.1.6
```

Note: I have a good ThinkPad P73, this is not the issue 😄

**Comment by williamdes** [comment]
Edit (v2):
```
Updating dependencies
Generating rules
Resolving dependencies through SAT
Looking at all rules.
Something's changed, looking at all rules again (pass #2)
Dependency resolution completed in 133.841 seconds
Analyzed 12698 packages to resolve dependencies
Analyzed 817234 rules to resolve dependencies
Nothing to modify in lock file
Generating rules
Resolving dependencies through SAT
Looking at all rules.

Dependency resolution completed in 0.001 seconds
Writing lock file
Installing dependencies from lock file (including require-dev)
Reading ./composer.lock (/mnt/Dev/@code-lts/Laravel-FCM/composer.lock)
Nothing to install, update or remove
Generating autoload files
75 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
```

**Comment by Seldaek** [comment]
Trying out with https://github.com/composer/composer/discussions/10276 I get this:

```
Updating dependencies with default package pool
Dependency resolution completed in 12.960 seconds
Analyzed 12706 packages to resolve dependencies
Analyzed 816991 rules to resolve dependencies

Updating dependencies with optimized package pool
Found 12,706 package versions referenced in your dependency graph. 10,584 (83%) were optimized away.
Dependency resolution completed in 0.832 seconds
Analyzed 2122 packages to resolve dependencies
Analyzed 96545 rules to resolve dependencies
Done, test successful
```

So I'm going to say it's resolved, or will be in 2.2 at least. Right now it works just a bit on the slow side.. Not sure how you get to 133seconds though as I consistently get results in 13seconds and my 4year old thinkpad isn't exactly a workhorse.

**Comment by williamdes** [comment]
Will try this and let you know, thanks :)
